qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java

```diff
@@ -186,15 +201,101 @@ private static void doCodeActionToDisableValidation(Template template, List<Diag
 		codeActions.add(disableValidationForTemplateQuickFix);
 	}
 
+	/**
+	 * Create CodeAction for deprecated `with` Qute syntax.
+	 *
+	 * @param template    the Qute template.
+	 * @param diagnostic  the diagnostic list that this CodeAction will fix.
+	 * @param codeActions the list of CodeActions to perform.
+	 * @throws BadLocationException
+	 *
+	 */
+	private void doCodeActionsForNotRecommendedWithSection(Template template, Diagnostic diagnostic,
+			List<CodeAction> codeActions) {
+		Range withSectionRange = diagnostic.getRange();
+		try {
+			// 1. Retrieve the #with section node using diagnostic range
+			int withSectionStart = template.offsetAt(withSectionRange.getStart());
+			WithSection withSection = (WithSection) template.findNodeAt(withSectionStart + 1);
+
+			// 2. Initialize TextEdit array
+			List<TextEdit> edits = new ArrayList<TextEdit>();
+
+			// 2.1 Create text edit to update section start from #with to #let
+			// and collect all object parts from expression for text edit
+			// (e.g. {name} -> {#let name=item.name})
+			edits.add(createWithSectionOpenEdit(template, withSection));
+
+			// 2.2 Create text edit to update section end from /with to /let
+			edits.add(createWithSectionCloseEdit(template, withSection));
+
+			// 3. Create CodeAction
+			CodeAction replaceWithSection = CodeActionFactory.replace(QUTE_DEPRICATED_WITH_SECTION, edits,
+					template.getTextDocument(), diagnostic);
+			codeActions.add(replaceWithSection);
+		} catch (BadLocationException e) {
+			return;
+		}
+	}
+
+	private TextEdit createWithSectionOpenEdit(Template template, WithSection withSection) throws BadLocationException {
+		String objectPartName = withSection.getObjectParameter() != null ? withSection.getObjectParameter().getName()
+				: "";
+
+		// Use set to avoid duplicate parameters
+		HashSet<String> withExpressions = new HashSet<String>();
+
+		// Retrieve all expressions in #with section
+		findExpressions(withSection, withExpressions);
+		List<String> letExpressionParameterList = new ArrayList<String>();
+		for (String expression : withExpressions) {
+			letExpressionParameterList.add(MessageFormat.format("{1}={0}.{1}", objectPartName, expression));
+		}
+
+		// Build text edit string
+		String letExpressionParameters = String.join(" ", letExpressionParameterList);
+		String withSectionOpenReplacementText = MessageFormat.format("#let {0}", letExpressionParameters);
+		Range withSectionOpen = new Range(template.positionAt(withSection.getStartTagNameOpenOffset()),
+				template.positionAt(withSection.getStartTagCloseOffset()));
+
+		return new TextEdit(withSectionOpen, withSectionOpenReplacementText);
+	}
+
+	private void findExpressions(Node node, HashSet<String> expressions) {
+		List<Node> children = node.getChildren();
+		// Base case: Node is an expression
+		if (children.isEmpty()) {
+			if (node.getKind() == NodeKind.Expression) {
+				expressions.add(((Expression) node).getContent());

```

Comment:
In case expression, you need to loop for each node of expressions to extract ExpressionPart (specially an Object Part).

I would use swithc(node.getKind()) to visit each nodes.

Reply:
Ypu need to takecar of Parameter of sectionwhich can have object part too.

Reply:
Is an ExpressionPart multiple Nodes? I was under the impression it is just one, i.e. `{name}, {price}, etc..`

Reply:
See sample at https://github.com/redhat-developer/quarkus-ls/blob/9c48330144a6eef07205ab4366f868a5bd753498/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java#L370

Reply:
I believe the Parameter use case should be handled now:
![replacewithwithletnestedparam](https://user-images.githubusercontent.com/26389510/157123755-3c8031db-59b5-4c1e-8ea8-cd6de0992028.gif)
I decided against a switch case however since we don't handle every SectionKind case for this code action

Reply:
@AlexXuChen I suggest that you rebase your PR from master to benefit with ASTVisitor. See https://github.com/redhat-developer/quarkus-ls/blob/b247849e376493117f3ab704d9f3060003ce86f0/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/parser/template/ASTVisitorTest.java#L59

I suggest that you copy/paste CollectObjectPartNameASTVisitor  (from test) in your code action class.

It will simplify the code and will cover other usecase that you don't take care of like `{item.name}`. In this case you must collect item and not item.name. The ASTVisitor takes care of that.